\documentclass[a4paper, english]{article}
\usepackage{graphics,eurosym,latexsym}
\usepackage{listings}
\lstset{columns=fixed,basicstyle=\ttfamily,numbers=left,numberstyle=\tiny,stepnumber=5,breaklines=true}
\usepackage{pst-all}
\usepackage{algorithmic,algorithm}
\usepackage{times}
\usepackage{babel}
\usepackage[nodayofweek]{datetime}
\usepackage[round]{natbib}
\bibliographystyle{plainnat}
\oddsidemargin=0cm
\evensidemargin=0cm
\textwidth=16cm
\textheight=23cm
\begin{document}

\title{\texttt{ms2dna} \input{version}: Convert Simulated Haplotypes to DNA Sequences}
\author{\input{author}}
\input{date}
\date{\displaydate{tagDate}}
\maketitle
\begin{abstract}
\texttt{ms2dna} is a program for converting haplotypes generated by the
popular coalescent simulation program \texttt{ms} to DNA sequences.
\end{abstract}

\section*{Introduction}
When testing DNA sequence analysis tools, it is often desirable to
simulate sequence samples that have evolved under well defined
scenarios. A simple method to achieve this is to convert the
haplotypes generated by available coalescent simulators to DNA
sequences. Perhaps the most widely used coalescent simulator is
\texttt{ms} and its user interface and output format have become de facto
standards~\cite{hud02:gen}. A fast alternative adhering to the same
format is \texttt{mspms}~\cite{kel16:eff}. Example output from \texttt{ms} is stored
in the file \texttt{data/simpleTest.dat} and looks like this:
\begin{verbatim}
ms 3 2 -t 10 -r 3 1000
52440 47412 34298

//
segsites: 4
positions: 0.1544 0.1855 0.2341 0.7006
1110
0110
0001

//
segsites: 7
positions: 0.2051 0.3380 0.3761 0.3982 0.4218 0.7364 0.8214
0000100
1111000
0000011
\end{verbatim}
The first line repeats the original command with which \texttt{ms} was
run. In this case it means that two samples of size three are
generated with neutral mutation parameter $\theta=4N\mu=10$ and
recombination parameter $R=4N\rho=3$ acting on a sequence of 1000
recombining loci or 1000 bp. The result consists of two sets of three
sequences. The first set has 4 segregating sites (mutations), the
second 7. These mutations are modeled as falling onto the interval
between 0 and 1, in the first data set they occur at positions
$0.1544, 0.1855, 0.2341, 0.7006$. Each row of zeros and ones
represents a haplotype, where a zero is the wild type and a 1 as the
mutant genotype.

The purpose of \texttt{ms2dna} is to convert the haplotypes generated by
\texttt{ms} into the corresponding DNA sequences of a certain length.
There already exists a freely available program that is often used for
this purpose: \texttt{seq-gen}~\cite{ram97:seq} takes as input a
genealogy, which might have been produced by \texttt{ms}, and generates a
compatible sequence sample. In contrast, \texttt{ms2dna} directly converts
the haplotypes generated by \texttt{ms} to DNA sequence data. This may be
advantageous in situations where the sample is generated by an
ancestral recombination graph rather than by a tree.

\section{Getting Started}
The program \texttt{ms2dna} was written in C on a computer running Linux.
Please contact \texttt{\input{email}} if there are any problems
with the program.
\begin{itemize}
\item Obtain the package\\
\texttt{git clone https://www.github.com/\input{repo}/ms2dna}
\item Change into the directory just downloaded
\begin{verbatim}
cd ms2dna
\end{verbatim}
and make \texttt{ms2dna}
\begin{verbatim}
make
\end{verbatim}
\item Test \texttt{ms2dna}
\begin{verbatim}
make test
\end{verbatim}
\item The executable \texttt{ms2dna} is located in the
  directory \texttt{build}. Place it into your \texttt{PATH}.
\item Make the documentation
\begin{verbatim}
make doc
\end{verbatim}
This calls the typesetting program \texttt{latex}, so please make sure
it is installed before making the documentation. The typeset manual is
located in
\begin{verbatim}
doc/ms2dna.pdf
\end{verbatim}
\end{itemize}

\section*{Program Details}
\subsection*{Input}
By default \texttt{ms2dna} reads the haplotype information in \texttt{ms}
format from the standard input (STDIN); it can also read it from one
or more files listed after the options. 

The program starts by reading parameters from the 
command line echoed by \texttt{ms}. In 
particular, it reads the length of the DNA sequence from the second
argument of the \texttt{-r} switch. For recombination-free data run \texttt{ms} with 0 as the
first argument of \texttt{-r}, e.g.
\begin{verbatim}
ms 3 1 -t 10 -r 0 1000 | ms2dna
\end{verbatim}
generates three recombination-free sequences of one kb length.

\subsection*{Output}
The output is printed in FASTA format
to the standard output stream (STDOUT). 
Each sample generated by \texttt{ms} is interpreted as representing data from an
independent locus. This is also reflected in the sequence headers where the
$y$-th sequence of the $x$-th locus would be identified as
\begin{verbatim}
>LX_SY
\end{verbatim}
The example data contained in the file \texttt{simpleTest.dat} consists of two
loci from which three sequences were sampled each.

The file \texttt{migrationTest.dat} contains simulated data drawn from a
structured population consisting of two subpopulations. For the interpretation of such 
data it is often convenient to include population information in the sequence
identifier. The $z$-th sequence from the $y$-th population at the $x$-th locus is
therefore labeled by \texttt{ms2dna}
\begin{verbatim}
>LX_PY_SZ
\end{verbatim}

\subsection*{Mutation Model}
\texttt{ms} implements the infinite sites model,
i. e. no position in a sequence can mutate twice. This is guaranteed by
representing the positions of mutations as random real numbers. 
With discrete sequences a given position in a DNA sequence may be hit
twice. In this case, \texttt{ms2dna} draws a new random
position. If the
number of segregating sites exceeds the number of mutable positions,
\texttt{ms2dna} prints a warning, e.g.:
\begin{verbatim}
#WARNING: number of segregating sites (1001) > number of mutable sites (1000)
\end{verbatim}
However, as long
as the sequence length is much larger than the expected number of segregating
sites, this should happen very rarely.

For each sample parsed by \texttt{ms2dna} a random ancestral sequence is
generated from scratch.
The ancestral sequence is then mutated at the positions output by \texttt{ms}.
The G/C content can be set using
the \texttt{-g} switch. Under the default G/C content (0.5), all nucleotides mutate with equal
probability into the other three nucleotides. If the $\textrm{G/C content} \ne
0.5$, this affects both the ancestral as well as in the mutated positions (as
expected).

\subsection*{Random Number Generator}
The sequences generated by \texttt{ms2dna} are, of course, random. As a result, if
\texttt{ms2dna} is run repeatedly on the same input data, different sequences are
generated every time. In order to guarantee this behavior, while still
allowing the user to switch it off conveniently, \texttt{ms2dna}
looks for a seed from two sources in the following order
\begin{itemize}
\item the argument to the \texttt{-s} switch; if this is not supplied (default),
\item the number contained the system clock plus the process ID.
\end{itemize}

\section{Change Log}
Please use
\begin{verbatim}
git log
\end{verbatim}
to list the change history.

\bibliography{ref}
\end{document}

